core-rust/core-api-server/src/core_api/handlers/stream_proofs.rs

```diff
@@ -0,0 +1,211 @@
+            )
+        }
+        StreamProofsFilterProtocolUpdateExecution {
+            from_state_version,
+            protocol_version,
+        } => iterate_protocol_update_execution_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_state_version(&database, from_state_version)?,
+            protocol_version,
+        ),
+    }?;
+
+    let (page, continuation_token) = to_api_page(
+        &mut proofs_iter,
+        page_size,
+        |proof| handlers::to_api_ledger_proof(&mapping_context, proof),
+        |proof| proof.ledger_header.state_version,
+    )?;
+
+    Ok(Json(models::StreamProofsResponse {
+        page,
+        continuation_token,
+    }))
+}
+
+fn iterate_all_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_state_version: StateVersion,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let start = optional_max(from_state_version, continue_from_state_version);
+
+    Ok(database.get_proof_iter(start))
+}
+
+fn iterate_end_of_epoch_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_epoch: Epoch,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let continuation_next_epoch = match continue_from_state_version {
+        Some(state_version) => match database.get_proof_iter(state_version).next() {
+            Some(proof) => Some(proof.ledger_header.epoch.next().unwrap()),
+            None => Err(client_error("continuation_token is not valid"))?,
+        },
+        None => None,
+    };
+
+    let start = optional_max(from_epoch, continuation_next_epoch);
+
+    Ok(database.get_next_epoch_proof_iter(start))
+}
+
+fn iterate_protocol_update_initialization_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_state_version: StateVersion,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let start = optional_max(from_state_version, continue_from_state_version);
+
+    Ok(database.get_protocol_update_init_proof_iter(start))
+}
+
+fn iterate_protocol_update_execution_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_state_version: StateVersion,
+    protocol_version: Option<String>,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let start = optional_max(from_state_version, continue_from_state_version);
+
+    let iter = database.get_protocol_update_execution_proof_iter(start);
+
+    Ok(match protocol_version {
+        Some(protocol_version) => Box::new(iter.filter(move |proof| {
```

Comment:
the `.filter()` here suggests that maybe we shouldn't try hard to support the protocol_version filter in the API?
_(unless it is a killer feature from convenience PoV)_

Reply:
There shouldn't be too many such proofs, so I don't think it's _too_ lacking in performance. I think it's a useful feature but if you're concerned about perf we could consider dropping it?

Reply:
> I think it's a useful feature

that's enough for me 👍